deps/aports/odrs.py

```diff
@@ -0,0 +1,103 @@
+import datetime
+import pathlib
+import plistlib
+import os
+import shutil
+import subprocess
+import ctypes
+
+if os.environ['ENABLE_ON_DEMAND_RESOURCES'] == 'YES':
+    raise Exception('This script collides with Xcode\'s ODR processing')
+
+def main():
+    root = pathlib.Path(os.environ['SRCROOT'])/'deps'/'aports'
+    repo_path = root/'main'/'x86'
+    packs = []
+    for file in repo_path.iterdir():
+        rsrc = ':'.join(file.relative_to(root).parts)
+        packs.append(([rsrc], {rsrc: str(file)}))
+    print('collected packs')
+    process_odrs(packs)
+
+# packs: [([tag], {resource_name: file_name}])]
+# OnDemandResources.plist: {
+#   NSBundleResourceRequestTags: {tag: [asset_pack]}
+#   NSBundleResourceRequestAssetPacks: {asset_pack: [filename]}
+# }
+# AssetPackManifestTemplate.plist: [{'bundleKey': asset_pack, 'URL': url}]
+def process_odrs(packs):
+    target_build_dir = pathlib.Path(os.environ['TARGET_BUILD_DIR'])
+    resources_dir = target_build_dir/os.environ['UNLOCALIZED_RESOURCES_FOLDER_PATH']
+    odrs_dir = target_build_dir/'OnDemandResources'
+    odrs_dir.mkdir(exist_ok=True)
+
+    packs = [(gen_pack_id(tags), tags, files) for (tags, files) in packs]
+    odr_tags_plist = {}
+    odr_packs_plist = {}
+    odrs_plist = {
+        'NSBundleResourceRequestTags': odr_tags_plist,
+        'NSBundleResourceRequestAssetPacks': odr_packs_plist,
+    }
+    packs_manifest_plist = {'resources': []}
+    for pack_id, tags, files in packs:
+        for tag in tags:
+            if tag not in odr_tags_plist:
+                odr_tags_plist[tag] = {'NSAssetPacks': []}
+            odr_tags_plist[tag]['NSAssetPacks'].append(pack_id)
+        odr_packs_plist[pack_id] = list(files.keys())
+
+        pack_path = odrs_dir/(pack_id+'.assetpack')
+        if pack_path.exists():
+            shutil.rmtree(pack_path)
+        pack_path.mkdir()
+        total_size = 0
+        with (pack_path/'Info.plist').open('wb') as f:
+            plistlib.dump({
+                'CFBundleIdentifier': pack_id,
+                'Tags': tags,
+            }, f, fmt=plistlib.FMT_BINARY)
```

Comment:
Would it be possible to use the XML format

Reply:
Is it better?

Reply:
It's readable

Reply:
Looks like this file gets shipped in the app store as-is, so we should make it as small as possible. It's 300k in binary form, and about 4x that in xml form.